streetwriters/notesnook
# Improve HTML parsing performance

This replaces DOMParser with htmlparser2 which is much, much faster. How much faster? 80%. This new implementation can parse at 50mb/s which is insane! The old one could only do 5-10mb/s

We still haven't gotten rid of the DOMParser though since HTML-to-MD conversion still needs it. This will be done soon though by using `dr-sax`.

This uses a custom implementation of htmlparser2 instead of the default one which is 50% faster.

**What will this affect?**

1. Saving — currently we need to parse the HTML on each edit to extract attachments etc. This slows down saving a lot especially if the note is big. Hopefully, this change will make saving instant even on really, really big notes.
2. Sync — we parse HTML of notes that might be conflicting to figure it whether to automatically resolve the conflict or not. This can be slow if the notes are really big. This change will improve the diffing performance to completely remove conflict resolution overhead.

**Testing:**

While the `core` test suite is passing, there is still potential for bugs since it all depends on the type of input being passed to htmlparser2. In my testing, it hasn't thrown any errors & saving has been consistently successful but we need to test it further:

1. Use all editor tools to create a diverse set of input including images, attachments, color, math etc.
2. Make sure saving is never interrupted.
3. Try this on mobile and see how it performs (cc @ammarahm-ed)
